core/src/cluster_info.rs

```diff
@@ -1755,37 +1773,94 @@ impl ClusterInfo {
         Some(packets)
     }
 
+    // Returns (failed, timeout, success)
     fn handle_pull_response(
         me: &Self,
         from: &Pubkey,
-        data: Vec<CrdsValue>,
+        mut crds_values: Vec<CrdsValue>,
         timeouts: &HashMap<Pubkey, u64>,
-    ) {
-        let len = data.len();
+    ) -> (usize, usize, usize) {
+        let len = crds_values.len();
         trace!("PullResponse me: {} from: {} len={}", me.id, from, len);
-        let (_fail, timeout_count) = me
+
+        if let Some(shred_version) = me.lookup_contact_info(from, |ci| ci.shred_version) {
+            Self::filter_by_shred_version(
+                from,
+                &mut crds_values,
+                shred_version,
+                me.my_shred_version(),
+            );
+        }
+        let filtered_len = crds_values.len();
+
+        let (fail, timeout_count, success) = me
             .time_gossip_write_lock("process_pull", &me.stats.process_pull_response)
-            .process_pull_response(from, timeouts, data, timestamp());
+            .process_pull_response(from, timeouts, crds_values, timestamp());
+
+        me.stats
+            .skip_pull_response_shred_version
+            .add_relaxed((len - filtered_len) as u64);
         me.stats.process_pull_response_count.add_relaxed(1);
-        me.stats.process_pull_response_len.add_relaxed(len as u64);
+        me.stats
+            .process_pull_response_len
+            .add_relaxed(filtered_len as u64);
         me.stats
             .process_pull_response_timeout
             .add_relaxed(timeout_count as u64);
+        me.stats.process_pull_response_fail.add_relaxed(fail as u64);
+        me.stats
+            .process_pull_response_success
+            .add_relaxed(success as u64);
+
+        (fail, timeout_count, success)
+    }
+
+    fn filter_by_shred_version(
+        from: &Pubkey,
+        crds_values: &mut Vec<CrdsValue>,
+        shred_version: u16,
+        my_shred_version: u16,
+    ) {
+        if shred_version != 0 && shred_version != my_shred_version {
+            // Allow someone to update their own ContactInfo if they
+            // can change shred versions if needed.
+            crds_values.retain(|crds_value| match &crds_value.data {
```

Comment:
clippy might yell at you to make this an `if let` since it's only matching one case.